File: Extensions/SciPy/cminpack-master/examples/chkdrv.f
c     **********
c
c     this program tests the ability of chkder to detect
c     inconsistencies between functions and their first derivatives.
c     fourteen test function vectors and jacobians are used. eleven of
c     the tests are false(f), i.e. there are inconsistencies between
c     the function vectors and the corresponding jacobians. three of
c     the tests are true(t), i.e. there are no inconsistencies. the
c     driver reads in data, calls chkder and prints out information
c     required by and received from chkder.
c
c     subprograms called
c
c       minpack supplied ... chkder,errjac,initpt,vecfcn
c
c     argonne national laboratory. minpack project. march 1980.
c     burton s. garbow, kenneth e. hillstrom, jorge j. more
c
c     **********
      integer i,ldfjac,lnp,mode,n,nprob,nread,nwrite
      integer na(14),np(14)
      logical a(14)
      double precision cp,one
      double precision diff(10),err(10),errmax(14),errmin(14),
     *                 fjac(10,10),fvec1(10),fvec2(10),x1(10),x2(10)
c
c     logical input unit is assumed to be number 5.
c     logical output unit is assumed to be number 6.
c
      data nread,nwrite /5,6/
c
      data a(1),a(2),a(3),a(4),a(5),a(6),a(7),a(8),a(9),a(10),a(11),
     *     a(12),a(13),a(14)
     *     /.false.,.false.,.false.,.true.,.false.,.false.,.false.,
     *      .true.,.false.,.false.,.false.,.false.,.true.,.false./
      data cp,one /1.23d-1,1.0d0/
      ldfjac = 10
   10 continue
         read (nread,60) nprob,n
         if (nprob .le. 0) go to 40
         call initpt(n,x1,nprob,one)
         do 20 i = 1, n
            x1(i) = x1(i) + cp
            cp = -cp
   20       continue
         write (nwrite,70) nprob,n,a(nprob)
         mode = 1
         call chkder(n,n,x1,fvec1,fjac,ldfjac,x2,fvec2,mode,err)
         mode = 2
         call vecfcn(n,x1,fvec1,nprob)
         call errjac(n,x1,fjac,ldfjac,nprob)
         call vecfcn(n,x2,fvec2,nprob)
         call chkder(n,n,x1,fvec1,fjac,ldfjac,x2,fvec2,mode,err)
         errmin(nprob) = err(1)
         errmax(nprob) = err(1)
         do 30 i = 1, n
            diff(i) = fvec2(i) - fvec1(i)
            if (errmin(nprob) .gt. err(i)) errmin(nprob) = err(i)
            if (errmax(nprob) .lt. err(i)) errmax(nprob) = err(i)
   30       continue
         np(nprob) = nprob
         lnp = nprob
         na(nprob) = n
         write (nwrite,80) (fvec1(i), i = 1, n)
         write (nwrite,90) (diff(i), i = 1, n)
         write (nwrite,100) (err(i), i = 1, n)
         go to 10
   40 continue
      write (nwrite,110) lnp
      write (nwrite,120)
      do 50 i = 1, lnp
         write (nwrite,130) np(i),na(i),a(i),errmin(i),errmax(i)
   50    continue
      stop
   60 format (2i5)
   70 format ( /// 5x, 8h problem, i5, 5x, 15h with dimension, i5, 2x,
     *         5h is  , l1)
   80 format ( // 5x, 25h first function vector    // (5x, 5d15.7))
   90 format ( // 5x, 27h function difference vector // (5x, 5d15.7))
  100 format ( // 5x, 13h error vector // (5x, 5d15.7))
  110 format (12h1summary of , i3, 16h tests of chkder /)
  120 format (46h nprob   n    status     errmin         errmax /)
  130 format (i4, i6, 6x, l1, 3x, 2d15.7)
c
c     last card of derivative check test driver.
c
      end

File: Extensions/SciPy/cminpack-master/examples/chkdrv.c
/* Usage: chkdrv < chkder.data */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "cminpack.h"
#include "vec.h"
#define real __cminpack_real__

/*     ********** */

/*     this program tests the ability of chkder to detect */
/*     inconsistencies between functions and their first derivatives. */
/*     fourteen test function vectors and jacobians are used. eleven of */
/*     the tests are false(f), i.e. there are inconsistencies between */
/*     the function vectors and the corresponding jacobians. three of */
/*     the tests are true(t), i.e. there are no inconsistencies. the */
/*     driver reads in data, calls chkder and prints out information */
/*     required by and received from chkder. */

/*     subprograms called */

/*       minpack supplied ... chkder,errjac,initpt,vecfcn */

/*     argonne national laboratory. minpack project. march 1980. */
/*     burton s. garbow, kenneth e. hillstrom, jorge j. more */

/*     ********** */

static void printvec(int n, const real *x)
{
    int i, num5, ilow, numleft;
    num5 = n/5;

    for (i = 0; i < num5; ++i) {
        ilow = i*5;
        printf("     %15.7e%15.7e%15.7e%15.7e%15.7e\n",
               (double)x[ilow+0], (double)x[ilow+1], (double)x[ilow+2], (double)x[ilow+3], (double)x[ilow+4]);
    }
    
    numleft = n%5;
    ilow = n - numleft;

    switch (numleft) {
        case 1:
            printf("     %15.7e\n",
                   (double)x[ilow+0]);
            break;
        case 2:
            printf("     %15.7e%15.7e\n",
                   (double)x[ilow+0], (double)x[ilow+1]);
            break;
        case 3:
            printf("     %15.7e%15.7e%15.7e\n",
                   (double)x[ilow+0], (double)x[ilow+1], (double)x[ilow+2]);
            break;
        case 4:
            printf("     %15.7e%15.7e%15.7e%15.7e\n",
                   (double)x[ilow+0], (double)x[ilow+1], (double)x[ilow+2], (double)x[ilow+3]);
            break;
    }
}

/* Main program */
int main(int argc, char **argv)
{
    /* Initialized data */
    const int a[14] = { 0,0,0,1,0,0,0,1,0,0,0,0,1,0 };
    real cp = .123;
    /* Local variables */
    int i, n;
    real x1[10], x2[10];
    int na[14], np[14];
    real err[10];
    int lnp = 0;
    real fjac[10*10];
    const int ldfjac = 10;
    real diff[10];
    real fvec1[10], fvec2[10];
    int nprob;
    real errmin[14], errmax[14];
    (void)argc; (void)argv;

    for (;;) {
        scanf("%5d%5d\n", &nprob, &n);
        if (nprob <= 0) {
            break;
        }

        hybipt(n,x1,nprob,1.);
        for(i=0; i<n; ++i) {
            x1[i] += cp;
            cp = -cp;
        }

        printf("\n\n\n      problem%5d      with dimension%5d   is  %c\n\n", nprob, n, a[nprob-1]?'T':'F');

        __cminpack_func__(chkder)(n,n,x1,NULL,NULL,ldfjac,x2,NULL,1,NULL);
        vecfcn(n,x1,fvec1,nprob);
        errjac(n,x1,fjac,ldfjac,nprob);
        vecfcn(n,x2,fvec2,nprob);
        __cminpack_func__(chkder)(n,n,x1,fvec1,fjac,ldfjac,NULL,fvec2,2,err);

        errmin[nprob-1] = err[0];
        errmax[nprob-1] = err[0];
        for(i=0; i<n; ++i) {
            diff[i] = fvec2[i] - fvec1[i];
            if (errmin[nprob-1] > err[i])
                errmin[nprob-1] = err[i];
            if (errmax[nprob-1] < err[i])
                errmax[nprob-1] = err[i];
        }

        np[nprob-1] = nprob;
        lnp = nprob;
        na[nprob-1] = n;

        printf("\n      first function vector   \n\n");
        printvec(n, fvec1);
        printf("\n\n      function difference vector\n\n");
        printvec(n, diff);
        printf("\n\n      error vector\n\n");
        printvec(n, err);
    }

    printf("\f summary of %3d tests of chkder\n", lnp);
    printf("\n nprob   n    status     errmin         errmax\n\n");

    for (i = 0; i < lnp; ++i) {
        printf("%4d%6d      %c   %15.7e%15.7e\n",
               np[i], na[i], a[i]?'T':'F', (double)errmin[i], (double)errmax[i]);
    }
    exit(0);
}
